internal/system/executor/docker.go

```diff
@@ -0,0 +1,138 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package executor
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+)
+
+const (
+	START   = "start"
+	STOP    = "stop"
+	RESTART = "restart"
+	METRICS = "metrics"
+	UNKNOWN = "unknown"
+	INSPECT = "inspect"
+
+	FAILED_START_PREFIX   = "Error starting service"
+	FAILED_RESTART_PREFIX = "Error restarting service"
+	FAILED_STOP_PREFIX    = "Error stopping service"
+)
+
+type CommandExecutor func(arg ...string) ([]byte, error)
+
+func Execute(operation string, service string, executor CommandExecutor) ([]byte, error) {
+	switch operation {
+	case START:
+		return nil, executeACommand(operation, service, executor, FAILED_START_PREFIX, true)
+	case RESTART:
+		return nil, executeACommand(operation, service, executor, FAILED_RESTART_PREFIX, true)
+	case STOP:
+		return nil, executeACommand(operation, service, executor, FAILED_STOP_PREFIX, false)
+	case METRICS:
+		return executor(
+			"stats",
+			service,
+			"--no-stream",
+			"--format",
+			"{\"cpu_perc\":\"{{ .CPUPerc }}\",\"mem_usage\":\"{{ .MemUsage }}\",\"mem_perc\":\"{{ .MemPerc }}\",\"net_io\":\"{{ .NetIO }}\",\"block_io\":\"{{ .BlockIO }}\",\"pids\":\"{{ .PIDs }}\"}")
+	default:
+		return nil, fmt.Errorf("operation not supported with specified executor")
+	}
+}
+
+func executorCommandFailedMessage(operationPrefix string, result string, errorMessage string) string {
+	return fmt.Sprintf("%s: %s (%s)", operationPrefix, errorMessage, result)
+}
+
+func executorCommandNotSupportedMessage() string {
+	return fmt.Sprintf("operation not supported with specified executor")
+}
+
+func executorInspectFailedMessage(operationPrefix string, errorMessage string) string {
+	return fmt.Sprintf("%s: %s", operationPrefix, errorMessage)
+}
+
+func serviceIsNotRunningButShouldBeMessage(operationPrefix string) string {
+	return fmt.Sprintf("%s: service is not running but should be", operationPrefix)
+}
+
+func serviceIsRunningButShouldNotBeMessage(operationPrefix string) string {
+	return fmt.Sprintf("%s: service is running but shouldn't be", operationPrefix)
+}
+
+func executeACommand(
+	operation string,
+	service string,
+	executor CommandExecutor,
+	operationPrefix string,
+	shouldBeRunning bool) error {
+
+	if result, err := executor(operation, service); err != nil {
+		return errors.New(executorCommandFailedMessage(operationPrefix, string(result), err.Error()))
```

Comment:
What you really want here are custom Error types rather than custom string building functions wrapped by the standard error type. We have [examples](https://github.com/edgexfoundry/edgex-go/blob/master/internal/core/data/errors/types.go) of this.

Reply:
Acknowledged.  Will review and incorporate in my changes.  Expect to see this in the new PR I intend to open.